ic-canister-client/src/pocket_ic.rs

```diff
@@ -0,0 +1,160 @@
+        Self {
+            client,
+            canister,
+            caller,
+        }
+    }
+
+    /// Returns the caller of the canister.
+    pub fn caller(&self) -> Principal {
+        self.caller
+    }
+
+    /// Replace the caller.
+    pub fn set_caller(&mut self, caller: Principal) {
+        self.caller = caller;
+    }
+
+    /// Returns the canister of the canister.
+    pub fn canister(&self) -> Principal {
+        self.canister
+    }
+
+    /// Replace the canister to call.
+    pub fn set_canister(&mut self, canister: Principal) {
+        self.canister = canister;
+    }
+
+    /// Returns the PocketIC client for the canister.
+    pub fn client(&self) -> &Arc<PocketIc> {
+        &self.client
+    }
+
+    /// Performs a blocking action with PocketIC client and awaits the result.
+    ///
+    /// Arguments of the closure `f`:
+    /// 1) `client` - The PocketIC client.
+    /// 2) `canister` - The canister principal.
+    /// 3) `caller` - The caller principal.
+    pub async fn with_client<F, R>(&self, f: F) -> R
+    where
+        F: Send + FnOnce(Arc<PocketIc>, Principal, Principal) -> R + 'static,
+        R: Send + 'static,
+    {
+        let client = self.client.clone();
+        let cansiter = self.canister;
+        let caller = self.caller;
+
+        tokio::task::spawn_blocking(move || f(client, cansiter, caller))
+            .await
+            .unwrap()
+    }
+
+    /// Performs update call with the given arguments.
+    pub async fn update<T, R>(&self, method: &str, args: T) -> CanisterClientResult<R>
+    where
+        T: ArgumentEncoder + Send + Sync,
+        R: for<'de> Deserialize<'de> + CandidType,
+    {
+        let args = candid::encode_args(args)?;
+        let method = String::from(method);
+
+        let call_result = self
+            .with_client(move |client, canister, caller| {
+                client.update_call(canister, caller, &method, args)
+            })
+            .await?;
+
+        let reply = match call_result {
+            WasmResult::Reply(reply) => reply,
+            WasmResult::Reject(e) => {
+                return Err(CanisterClientError::CanisterError((
+                    RejectionCode::CanisterError,
+                    e,
+                )));
+            }
+        };
+
+        let decoded = Decode!(&reply, R)?;
+        Ok(decoded)
+    }
+
+    /// Performs query call with the given arguments.
+    pub async fn query<T, R>(&self, method: &str, args: T) -> CanisterClientResult<R>
+    where
+        T: ArgumentEncoder + Send + Sync,
+        R: for<'de> Deserialize<'de> + CandidType,
+    {
+        let args = candid::encode_args(args)?;
+        let method = String::from(method);
+
+        let call_result = self
+            .with_client(move |env, canister, caller| {
+                env.query_call(canister, caller, &method, args)
+            })
+            .await?;
+
+        let reply = match call_result {
```

Comment:
Looks like error conversion can be extracted to a function to avoid duplication

Reply:
Done